File: python/src/threads/team/rebuild.py
"""Rebuilding a team's index from its logs alone (the replay rule): the lead's log, every member
log and the team log, read verified, checked across (rule 43), then folded with the same
`insert_rows` / `change_rows` the writer uses, in one transaction."""

import sqlite3
from collections.abc import Sequence
from dataclasses import dataclass
from typing import TYPE_CHECKING

from pydantic.experimental.missing_sentinel import MISSING

from threads.log import ParseError, TeamOpenedEvent, ThreadId, ThreadStartedEvent
from threads.result import Err, Ok
from threads.store import wakes
from threads.store.deletion import TEAM_TABLES
from threads.store.sql import export, transaction
from threads.store.started import Opened, opened_threads
from threads.store.verify import VerifiedLog, verify_export
from threads.team.cross import TeamLogEvents, check_team_logs
from threads.team.index import TeamLog, change_rows, insert_rows, turn_openers

if TYPE_CHECKING:
    from threads.store import SqliteStore


@dataclass(frozen=True, slots=True)
class _Read:
    log: TeamLog
    verified: VerifiedLog


def team_branches(conn: sqlite3.Connection, tenant_id: str, team_id: str) -> list[TeamLog]:
    """The team's logs, by branch: its lead's, every member's (parent team_member through the
    lead) and its team log; empty when no lead names the team."""
    opened = opened_threads(conn, tenant_id)
    lead = next((o for o in opened if _leads(o, team_id)), None)
    if lead is None:
        return []
    logs = [
        TeamLog(o.thread_id, o.branch_id)
        for o in opened
        if o is lead or _in_team(o, lead.thread_id, team_id)
    ]
    return sorted(logs, key=lambda log: log.branch_id)


def _leads(o: Opened, team_id: str) -> bool:
    e = o.event
    return (
        isinstance(e, ThreadStartedEvent)
        and e.data.team is not MISSING
        and e.data.team.id == team_id
    )


def _in_team(o: Opened, lead: ThreadId, team_id: str) -> bool:
    """A member (parent team_member through the lead) or the team log."""
    e = o.event
    if isinstance(e, TeamOpenedEvent):
        return e.data.team == team_id
    parent = e.data.parent
    return parent is not MISSING and parent.relation == "team_member" and parent.thread_id == lead


async def rebuild_team_index(store: "SqliteStore", team_id: str) -> Ok[None] | Err[ParseError]:
    """Wipes and refolds every index row of `team_id` (and its branches' pending_wakes); the
    feed starts a new epoch. not_found: no lead names the team. invalid_transition: rule 43. The
    reads, the check and the refold are one transaction, so no append lands between the read
    and the wipe."""
    tenant = store.tables.tenant_id
    return await store.run(lambda c: _rebuild(c, tenant, team_id))


def _rebuild(conn: sqlite3.Connection, tenant: str, team_id: str) -> Ok[None] | Err[ParseError]:
    with transaction(conn):
        # Nothing is written before a refusal, so the transaction commits nothing then.
        return refold_team(conn, tenant, team_id)


def refold_team(conn: sqlite3.Connection, tenant: str, team_id: str) -> Ok[None] | Err[ParseError]:
    """`rebuild_team_index` in the caller's transaction (an import's)."""
    logs = team_branches(conn, tenant, team_id)
    if not logs:
        return Err(ParseError("not_found", f"no team {team_id}"))
    reads = _read_all(conn, logs)
    if isinstance(reads, Err):
        return reads
    broken = check_team_logs(
        [TeamLogEvents(r.log.thread_id, r.log.branch_id, r.verified.fold.events) for r in reads],
        team_id,
    )
    if broken is not None:
        message = f"{broken.branch_id}: {broken.message}"
        return Err(ParseError("invalid_transition", message, broken.seq))
    _refold(conn, team_id, reads)
    return Ok(None)


def _read_all(conn: sqlite3.Connection, logs: Sequence[TeamLog]) -> list[_Read] | Err[ParseError]:
    """Each log read back through the store's boundary, on the rebuild's own transaction."""
    reads: list[_Read] = []
    for log in logs:
        read = verify_export(export(conn, log.branch_id), 0)
        if isinstance(read, Err):
            error = read.error
            return Err(ParseError(error.code, f"{log.branch_id}: {error.message}", error.seq))
        reads.append(_Read(log, read.value))
    return reads


def _refold(conn: sqlite3.Connection, team_id: str, reads: Sequence[_Read]) -> None:
    (epoch,) = conn.execute(
        "SELECT COALESCE(MAX(epoch), 0) FROM team_feed WHERE team_id = ?", (team_id,)
    ).fetchone()
    for table in TEAM_TABLES:
        conn.execute(f"DELETE FROM {table} WHERE team_id = ?", (team_id,))  # noqa: S608
    for r in reads:
        wakes.refold(conn, r.log.branch_id, r.verified.fold.events)
    for r in reads:
        insert_rows(conn, r.log, r.verified.fold.events, team_id)
    for r in reads:
        change_rows(conn, r.log, r.verified.fold.events, turn_openers(r.verified), team_id)
    feed = sorted(
        (r.log.branch_id, e.seq) for r in reads for e, _ in r.verified.segments[-1].events
    )
    conn.executemany(
        "INSERT INTO team_feed (team_id, epoch, feed_offset, branch_id, seq)"
        " VALUES (?, ?, ?, ?, ?)",
        [(team_id, int(epoch) + 1, i, b, s) for i, (b, s) in enumerate(feed, 1)],
    )

File: typescript/packages/core/src/team/rebuild.ts
import { z } from "zod";
import type { TeamId } from "../log";
import { knownEvents } from "../reduce";
import { err, ok, type Result } from "../result";
import type { LogStore } from "../store";
import { TEAM_TABLES } from "../store/deletion";
import { READ_ONLY, type Tx } from "../store/driver";
import { openedThreads } from "../store/started";
import { atomically, parseRows } from "../store/tables";
import { refoldWakes } from "../store/wakes";
import type { VerifiedLog } from "../verify";
import { type LogError, logError } from "../verify/error";
import { checkTeamLogs } from "./cross";
import { changeRows, insertRows, type TeamLog, turnOpeners } from "./index";

// The replay rule's other half: wipe a team's index rows and fold them again from its logs
// alone, with the same insertRows/changeRows a writer uses, so a rebuild equals what the appends
// wrote (mail's claim columns aside, and team_feed, which starts a new epoch).

/** One of the team's logs, verified. */
export type TeamChain = TeamLog & { readonly chain: VerifiedLog };

/**
 * The team's logs in `store`'s tenant, each main branch read verified, in branch order: the
 * lead's (its thread_started names the team), its members' (parent relation team_member naming
 * the lead's thread) and the team log. not_found when no lead names the team.
 */
export function teamChains(
  store: LogStore,
  teamId: TeamId,
): Promise<Result<readonly TeamChain[], LogError>> {
  return store.driver.transaction(
    (tx) => chainsIn(tx, store, teamId),
    READ_ONLY,
  );
}

async function chainsIn(
  tx: Tx,
  store: LogStore,
  teamId: TeamId,
): Promise<Result<readonly TeamChain[], LogError>> {
  const opened = await openedThreads(tx, store.tenant);
  if (!opened.ok) return opened;
  const lead = opened.value.find(
    (o) =>
      o.event.type === "thread_started" && o.event.data.team?.id === teamId,
  );
  if (lead === undefined || lead.event.type !== "thread_started")
    return err(logError("not_found", `no lead of team ${teamId}`));
  const logBranch = lead.event.data.team?.log_branch_id;
  const members = opened.value.filter(
    (o) =>
      o.event.type === "thread_started" &&
      o.event.data.parent?.relation === "team_member" &&
      o.event.data.parent.thread_id === lead.threadId,
  );
  const teamLog = opened.value.filter(
    (o) => o.event.type === "team_opened" && o.branchId === logBranch,
  );
  const chains: TeamChain[] = [];
  for (const o of [lead, ...members, ...teamLog]) {
    const chain = await store.readIn(tx, o.branchId);
    if (!chain.ok) return chain;
    chains.push({
      threadId: o.threadId,
      branchId: o.branchId,
      chain: chain.value,
    });
  }
  return ok(chains.toSorted((a, b) => byText(a.branchId, b.branchId)));
}

/**
 * In one IMMEDIATE transaction, so no append lands between the read and the refold: reads the
 * team's logs, checks rule 43 across them, wipes every index row of the team (and the wake rows
 * of its branches) and folds them again: every log's inserts, then every log's changes, then the
 * feed under a new epoch, offsets in (branch_id, seq) order.
 */
export function rebuildTeamIndex(
  store: LogStore,
  teamId: TeamId,
  outer?: Tx,
): Promise<Result<void, LogError>> {
  return atomically(outer ?? store.driver, async (tx) => {
    const chains = await chainsIn(tx, store, teamId);
    if (!chains.ok) return chains;
    const logs = chains.value.map((c) => ({
      ...c,
      events: knownEvents(c.chain),
    }));
    const broken = checkTeamLogs(logs, teamId);
    if (broken !== undefined)
      return err(
        logError(
          "invalid_transition",
          `branch ${broken.branchId}: ${broken.message}`,
          broken.seq,
        ),
      );
    const epoch = await nextEpoch(tx, teamId);
    if (!epoch.ok) return epoch;
    await wipe(tx, teamId);
    for (const log of logs) await refoldWakes(tx, log.branchId, log.events);
    for (const log of logs) await insertRows(tx, log, log.events, teamId);
    for (const log of logs)
      await changeRows(
        tx,
        log,
        log.events,
        turnOpeners(log.chain.events),
        teamId,
      );
    const feed = logs
      .flatMap((log) => log.events.map((e) => [log.branchId, e.seq] as const))
      .toSorted(([a, x], [b, y]) => byText(a, b) || x - y);
    for (const [i, [branch, seq]] of feed.entries())
      await tx.run(
        "INSERT INTO team_feed (team_id, epoch, feed_offset, branch_id, seq) VALUES (?, ?, ?, ?, ?)",
        [teamId, epoch.value, i + 1, branch, seq],
      );
    return ok(undefined);
  });
}

async function wipe(tx: Tx, teamId: TeamId): Promise<void> {
  for (const table of TEAM_TABLES)
    await tx.run(`DELETE FROM ${table} WHERE team_id = ?`, [teamId]);
}

/** Code-unit order, as SQLite's BINARY collation and Python sort ids. */
const byText = (a: string, b: string): number => (a < b ? -1 : a > b ? 1 : 0);

const Epoch = z.strictObject({ epoch: z.int().nullable() });

/** A rebuilt feed starts a new epoch, so a cursor from an older one restarts. */
async function nextEpoch(
  tx: Tx,
  teamId: TeamId,
): Promise<Result<number, LogError>> {
  const rows = parseRows(
    Epoch,
    await tx.all(
      "SELECT MAX(epoch) AS epoch FROM team_feed WHERE team_id = ?",
      [teamId],
    ),
  );
  return rows.ok ? ok((rows.value[0]?.epoch ?? 0) + 1) : rows;
}
